File: src/main/java/co/uk/isxander/evergreenhud/elements/impl/ElementYaw.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.elements.impl;

import co.uk.isxander.evergreenhud.elements.Element;
import co.uk.isxander.evergreenhud.settings.impl.BooleanSetting;
import co.uk.isxander.evergreenhud.elements.ElementData;
import net.minecraft.util.MathHelper;

import java.text.DecimalFormat;

public class ElementYaw extends Element {

    public BooleanSetting trailingZeros;

    @Override
    public void initialise() {
        addSettings(trailingZeros = new BooleanSetting("Trailing Zeros", "Display", "Add zeroes to match the accuracy.", false));
    }

    @Override
    public ElementData metadata() {
        return new ElementData("Yaw Display", "Shows the player's rotation yaw. Useful for bridging.");
    }

    @Override
    protected String getValue() {
        return new DecimalFormat(trailingZeros.get() ? "0.0" : "#.#").format(MathHelper.wrapAngleTo180_float(mc.thePlayer.rotationYaw));
    }

    @Override
    public String getDefaultDisplayTitle() {
        return "Yaw";
    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementYaw.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.impl

import dev.isxander.evergreenhud.elements.type.SimpleTextElement
import dev.isxander.evergreenhud.utils.elementmeta.ElementMeta
import dev.isxander.evergreenhud.utils.mc
import dev.isxander.settxi.impl.boolean
import dev.isxander.settxi.impl.int
import net.minecraft.util.MathHelper
import java.text.DecimalFormat

@ElementMeta(id = "evergreenhud:yaw", name = "Yaw", description = "Displays the player's yaw.", category = "Player")
class ElementYaw : SimpleTextElement("Yaw", 0) {
    var accuracy by int(2) {
        range = 0..8
        name = "Accuracy"
        category = "Speed"
        description = "How many decimal places to show the speed to."
    }

    var trailingZeros by boolean(true) {
        name = "Trailing Zeros"
        category = "Speed"
        description = "Show zeros to match the accuracy."
    }

    override fun calculateValue(): String {
        val format = if (trailingZeros) "0" else "#"

        return DecimalFormat("0${if (accuracy > 0) "." else ""}" + format.repeat(accuracy)).format(MathHelper.wrapAngleTo180_float(mc.thePlayer.rotationYaw))
    }
}
